src/proc/constant_evaluator.rs

```diff
@@ -62,169 +75,109 @@ pub enum ConstantSolvingError {
     SwizzleVectorOnly,
     #[error("Type is not constructible")]
     TypeNotConstructible,
+    #[error("Subexpression(s) are not constant")]
+    SubexpressionsAreNotConstant,
     #[error("Not implemented as constant expression: {0}")]
     NotImplemented(String),
 }
 
-#[derive(Clone, Copy)]
-pub enum ExprType {
-    Regular,
-    Constant,
+// Access
+// AccessIndex
+// Splat
+// Swizzle
+// Unary
+// Binary
+// Select
+// Relational
+// Math
+// As
+
+// TODO(teoxoy): consider accumulating this metadata instead of recursing through subexpressions
```

Comment:
Please don't leave TODO comments in the code. Instead, file PRs for follow-up work, and link to them in the code if that's valuable.

Reply:
Done in https://github.com/gfx-rs/naga/pull/2309/commits/0d536b998dbfe2eb5a14167a365ba4fbee25d02d